### PR TITLE
Split batch and retry if message_too_large

### PR DESCRIPTION
**Pull request by zmstone**
When message_too_large is received from Kafka, the producer will:
1. If single-message is too large, drop the message, and reply error to caller, also log a `error` level message
2. For `size>=2` batches, split the batches into single-message batches, and retry, then lower the config `max_batch_bytes` to half.

### CODE REVIEW

**Review by terry-xiaoyu — APPROVED** [review]

File: src/wolff_producer.erl [hunk 1]
```diff
@@ -446,35 +505,81 @@ handle_kafka_ack(#kpro_rsp{api = produce,
...
-         };
+      %% this clause is kept only to be hot-upgrade safe
+      clear_sent_and_ack_callers(Q_AckRef, Calls, BaseOffset, St);
+    {value, ?sent_items(Ref, Items, Q_AckRef)} when Reason =:= ?message_too_large ->
+      Batch = get_batch_from_queue_items(Items),
+      case length(Batch) of
+        1 ->
+          %% This is a single message batch, but it's still too large
+          #{topic := Topic, partition := Partition} = St,
+          log_error(Topic, Partition, "One message is dropped due to single-message batch is too large!", []),
```
Comment (thalesmg) [inline src/wolff_producer.erl:530]: Would it be worth throttling this kind of message, similar to what is done when logging discarded messages?

File: src/wolff_producer.erl [hunk 2]
```diff
@@ -446,35 +505,81 @@ handle_kafka_ack(#kpro_rsp{api = produce,
...
+          Calls = get_calls_from_queue_items(Items),
+          clear_sent_and_ack_callers(Q_AckRef, Calls, ?message_too_large, St);
+        _ ->
+          #{topic := Topic,
+            partition := Partition,
+            config := #{max_batch_bytes := Max} = Config
+           } = St,
+          NewMax = (Max + 1) div 2,
```
Comment (thalesmg) [inline src/wolff_producer.erl:538]:
```suggestion
          NewMax = max(1, (Max + 1) div 2),
```

File: test/wolff_tests.erl [hunk 3]
```diff
@@ -460,10 +438,91 @@ test_leader_restart() ->
...
+    ClientCfg = client_config(),
+    ClientId = iolist_to_binary("client-" ++ Topic),
+    {ok, Client} = start_client(ClientId, ?HOSTS, ClientCfg#{connection_strategy => per_partition}),
+    TopicBin = iolist_to_binary(Topic),
+    %% try to batch more messages than Kafka's limt,
+    %% the producer will get message_too_large error back
+    %% then it should retry sending one message at a time
+    ProducerCfg = #{partitioner => fun(_, _) -> 0 end, max_batch_bytes => MaxMessageBytes * 3},
+    {ok, Producers} = wolff:start_producers(Client, TopicBin, ProducerCfg),
+    MaxBytesCompensateOverhad = MaxMessageBytes - ?BATCHING_OVERHEAD - 7,
```
Comment (thalesmg) [inline test/wolff_tests.erl:473]:
```suggestion
    MaxBytesCompensateOverhead = MaxMessageBytes - ?BATCHING_OVERHEAD - 7,
```

File: test/wolff_tests.erl [hunk 4]
```diff
@@ -460,10 +438,91 @@ test_leader_restart() ->
...
+  Topic = "message-too-large-" ++ integer_to_list(abs(erlang:monotonic_time())),
+  Partitions = 1,
+  ReplicationFactor = 1,
+  MaxMessageBytes = 100,
+  with_topic(Topic, Partitions, ReplicationFactor, MaxMessageBytes, fun() ->
+    ClientCfg = client_config(),
+    ClientId = iolist_to_binary("client-" ++ Topic),
+    {ok, Client} = start_client(ClientId, ?HOSTS, ClientCfg#{connection_strategy => per_partition}),
+    TopicBin = iolist_to_binary(Topic),
+    %% try to batch more messages than Kafka's limt,
```
Comment (thalesmg) [inline test/wolff_tests.erl:468]:
```suggestion
    %% try to batch more messages than Kafka's limit,
```

File: test/wolff_tests.erl [hunk 5]
```diff
@@ -460,10 +438,91 @@ test_leader_restart() ->
...
+one_byte_limit_test() ->
+  Topic = "one-byte-limit-" ++ integer_to_list(abs(erlang:monotonic_time())),
+  Partitions = 1,
+  ReplicationFactor = 1,
+  with_topic(Topic, Partitions, ReplicationFactor, _MessageSizeLimit = 1, fun() ->
+    ClientCfg = client_config(),
+    ClientId = iolist_to_binary("client-" ++ Topic),
+    {ok, Client} = start_client(ClientId, ?HOSTS, ClientCfg#{connection_strategy => per_partition}),
+    TopicBin = iolist_to_binary(Topic),
+    %% try to batch more messages than Kafka's limt,
```
Comment (thalesmg) [inline test/wolff_tests.erl:505]:
```suggestion
    %% try to batch more messages than Kafka's limit,
```

**Review by qzhuyan — APPROVED** [review]

**Review by thalesmg — APPROVED** [review]

File: src/wolff_producer.erl (hunk 1 above)
Comment (zmstone) [reply src/wolff_producer.erl:530]: the intention is to be very noisy

**Review by thalesmg — APPROVED** [review]

**Pull request merged**